services/inputhost/pathCache.go

```diff
@@ -508,3 +536,26 @@ func (pathCache *inPathCache) getState() *admin.DestinationState {
 
 	return destState
 }
+
+func (pathCache *inPathCache) drainExtent(extUUID string, updateUUID string, drainWG *sync.WaitGroup) {
+	defer drainWG.Done()
+	pathCache.RLock()
+	defer pathCache.RUnlock()
+
+	extCache, ok := pathCache.extentCache[extentUUID(extUUID)]
+	if !ok {
+		// draining called for extent which is not present
+		// just return
+		return
+	}
+
+	if extCache.connection.prepDrain() {
+		var connDrainWG sync.WaitGroup
+		// first send DRAIN command to all connections
+		// then start draining extents
+		pathCache.drainConnections(updateUUID, &connDrainWG)
+		// do best effort waiting to notify all connections here
+		common.AwaitWaitGroup(&connDrainWG, connWGTimeout)
+		extCache.connection.drain()
```

Comment:
Even after we send the DRAIN message to the client and all the connections have drained (i.e, they have flushed their buffers which means only the *write* side is done), if we do a simple close here, the exthost layer (which is buffered as well) will fail any inflight messages it has in it's buffer where we could do better and get ACKs from stores as well. Also this the reason i don't use the entire drain timeout for the connection WG but rather a few seconds to give the clients a chance to flush on their side.